examples/classic/distributed/TwoPhaseCommit/two_phase_commit.qnt

```diff
@@ -0,0 +1,162 @@
+// -*- mode: Bluespec; -*-
+
+/* This specification stems from the specification of the two phase protocol
+   by Leslie Lamport:
+
+   https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/TwoPhase.tla */
+
+module two_phase_commit {
+  type ResourceManagerState = Working | Prepared | Committed | Aborted
+  var resourceManagerStates: str -> ResourceManagerState
+
+  type TransactionManagerState = Init | Done
+  var transactionManagerState: TransactionManagerState
+
+  /// The set of resource managers from which the transaction manager has
+  /// received RMPrepared messages
+  var preparedResourceManagers: Set[str]
+
+  type Message =
+    | Abort
+    | Commit
+    | RMPrepared(str)
+    | RMCommitted(str)
+
+  var messages: Set[Message]
+
+  const resourceManagers: Set[str]
+
+  action init = all {
+    resourceManagerStates' = resourceManagers.mapBy(r => Working),
+    transactionManagerState' = Init,
+    preparedResourceManagers' = Set(),
+    messages' = Set(),
+  }
+
+  /* Transaction Manager actions */
+
+  /// The transaction manager receives a RMPrepared(rm) message from the resource manager rm
+  action addPreparedResourceManager(resourceManager: str): bool = all {
+    transactionManagerState == Init,
+    messages.contains(RMPrepared(resourceManager)),
+    preparedResourceManagers' = preparedResourceManagers.union(Set(resourceManager))
+  }
```

Comment:
> With the way the action is written here, it leaves open the possibility that the action could be conjoined with, e.g., prepare, and both could occur simultaneously.

I don't think I have thought about this problem before! This is interesting.

> Are both specs correct ways of modelling the system?
I'd say yes because the only action that really matters is `step`, and that one doesn't combine this with, e.g., `prepare`. This is a matter of spec organization and preventing mistakes while writing, I think. I'll talk more about this in the reply for your other question on `abort`